### PR TITLE
Python: support Azure AD auth

### PR DESCRIPTION
**Pull request by jjhenkel**
### Motivation and Context

We've been wanting to add support for AD auth for the `Azure*` backends, this adds that.

### Description

Backends like `AzureTextCompletion` can be instantiated in two ways now. Either: 
```
AzureTextCompletion(deployment_name, endpoint, api_key)
```

Or (optionally, you could also manually set endpoint and key):
```
AzureTextCompletion(deployment_name, use_ad_auth=True)
```

In some contexts (like in notebooks running on Azure) we can get endpoint/key automagically. In other cases, you can now use AD auth and explicitly pass the right key/endpoint.

### Contribution Checklist

image

- [x] The code builds clean without any errors or warnings
- [x] The PR follows SK Contribution Guidelines (https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
- [x] The code follows the .NET coding conventions (https://learn.microsoft.com/dotnet/csharp/fundamentals/coding-style/coding-conventions) verified with `dotnet format`
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anything :smile:

### CODE REVIEW

**Review by awharrison-28 — CHANGES_REQUESTED** [review]
The README and .env.example is missing AZURE_OPENAI_DEPLOYMENT_NAME=""

**Comment by jjhenkel** [comment]
Thanks @awharrison-28! Missing example key added to `.env.example`

**Review by awharrison-28 — CHANGES_REQUESTED** [review]

File: python/semantic_kernel/utils/notebooks.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+
+def try_get_api_info_from_synapse_mlflow() -> Optional[Tuple[str, str]]:
+    """
+    Attempts to get the endpoint and api_key from the Synapse MLFlow environment.
+
+    :return: The endpoint and api_key (or None, if not found)
+    """
+    try:
+        from synapse.ml.mlflow import get_mlflow_env_config
```
[the remark this answers is not in the document]
Comment (awharrison-28) [reply python/semantic_kernel/utils/notebooks.py:13]: Returning None still I fine I think. An extra message would be useful though.

[the remark this answers is not in the document]
Comment (jjhenkel) [reply python/semantic_kernel/utils/notebooks.py:13]: I added error handling to the `Azure*` backends to call out that you need to pass a key/endpoint (this is true for pretty much every context except special environments, so I think it's a good error).

Also added tests for the constructors/init logic!

File: python/semantic_kernel/ai/open_ai/services/azure_chat_completion.py [hunk 2]
```diff
@@ -7,38 +7,53 @@
...
         logger: Optional[Logger] = None,
+        use_ad_auth=False,
     ) -> None:
+        if use_ad_auth and (endpoint is None or api_key is None):
+            # Try to get endpoint/api_key via alternative methods
+            results = try_get_api_info_from_synapse_mlflow()
+            if results is not None:
+                endpoint, api_key = results
+            else:
+                raise ValueError(
```
Comment (awharrison-28) [inline python/semantic_kernel/ai/open_ai/services/azure_chat_completion.py:33]: suggestion: raise ValueError("Azure AD authentication failed. Alternatively provide an endpoint and API key.")

Comment (jjhenkel) [reply python/semantic_kernel/ai/open_ai/services/azure_chat_completion.py:33]: Error message and tests updated

File: python/semantic_kernel/utils/notebooks.py
Comment (dluc) [inline python/semantic_kernel/utils/notebooks.py:1]: The SDK config pattern is about passing all the settings one by one, e.g. endpoint URLs and API keys. This approach is instead hardcoding a specific config provider telling the kernel to use it.

The code needs be refactored, moving this function outside the kernel, allowing users to pass in AAD endpoint and AAD tokens, when they need to, e.g. the kernel doesn't need to know whether there is a Synapse MLFlow environment, similarly to how the kernel doesn't know about Azure KeyVault, .NET Secret Manager, environment variables, etc.

**Review by dluc — CHANGES_REQUESTED** [review]
try_get_api_info_from_synapse_mlflow to be refactored out, allowing AAD auth details to be passed in similarly to other settings

**Comment by jjhenkel** [comment]
Sure, there's no technical reason it can't be factored out. It's all about convenience and seamless experience. You write 5 lines to set this up, or you write 1. (In notebooks contexts.)

My only technical goal here is enabling AAD auth: this PR does that. My usability/ease-of-use goal is making setup in notebooks environments as easy as possible. There's no technical merit to that. There's no technical blockers to refactoring. It's simply a "you could do this in 1 line or 5" choice.

**Comment by jjhenkel** [comment]
Compromise implemented: we do dynamic loading for 3rd-party auth. (But, we add a bit of generic code to the kernel to pick up on 3rd-party auth providers.)

**Comment by dluc** [comment]
Here's what should suffice to support AAD, without the need to introduce utils and auth providers:

```python
# Copyright (c) Microsoft. All rights reserved.


from logging import Logger
from typing import Any, Optional

from semantic_kernel.ai.open_ai.services.open_ai_chat_completion import (
    OpenAIChatCompletion,
)


class AzureChatCompletion(OpenAIChatCompletion):
    _endpoint: str
    _api_version: str
    _api_type: str

    def __init__(
        self,
        deployment_name: str,
        endpoint: Optional[str] = None,
        api_key: Optional[str] = None,
        use_aad_auth = False,
        api_version: str = "2023-03-15-preview",
        logger: Optional[Logger] = None,
    ) -> None:
        """
        Initialize an AzureChatCompletion backend.

        :param deployment_name: The name of the Azure deployment. This value
            will correspond to the custom name you chose for your deployment
            when you deployed a model. This value can be found under
            Resource Management > Deployments in the Azure portal or, alternatively,
            under Management > Deployments in Azure OpenAI Studio.
        :param endpoint: The endpoint of the Azure deployment. This value
            can be found in the Keys & Endpoint section when examining
            your resource from the Azure portal.
        :param api_key: The API key for the Azure deployment. This value can be
            found in the Keys & Endpoint section when examining your resource in
            the Azure portal. You can use either KEY1 or KEY2. If use_ad_auth is
            True, use this parameter to pass the AAD token.
        :param use_aad_auth: Whether to use Azure Active Directory authentication.
            (Optional) The default value is False.
        :param api_version: The API version to use. (Optional)
            The default value is "2022-12-01".
        :param logger: The logger instance to use. (Optional)
        """

        if not deployment_name:
            raise ValueError("The deployment name cannot be `None` or empty")
        if not api_key:
            raise ValueError("The Azure API key cannot be `None` or empty`")
        if not endpoint:
            raise ValueError("The Azure endpoint cannot be `None` or empty")
        if not endpoint.startswith("https://"):
            raise ValueError("The Azure endpoint must start with https://")

        self._endpoint = endpoint
        self._api_version = api_version
        self._api_type = "azure_ad" if use_aad_auth else "azure"

        super().__init__(deployment_name, api_key, org_id=None, log=logger)

    def _setup_open_ai(self) -> Any:
        import openai

        openai.api_type = self._api_type
        openai.api_key = self._api_key
        openai.api_base = self._endpoint
        openai.api_version = self._api_version

        return openai
```

**Comment by jjhenkel** [comment]
Removed `auth_providers` (and related tests) --- should be ready to merge now

File: python/semantic_kernel/ai/open_ai/services/azure_chat_completion.py [hunk 3]
```diff
@@ -12,33 +12,59 @@
...
+            (Optional) The default value is False.
+        """
         if not deployment_name:
             raise ValueError("The deployment name cannot be `None` or empty")
         if not api_key:
             raise ValueError("The Azure API key cannot be `None` or empty`")
         if not endpoint:
             raise ValueError("The Azure endpoint cannot be `None` or empty")
-        if not endpoint.startswith("https://"):
+        if not endpoint.startswith("https://") and not use_ad_auth:
```
Comment (dluc) [inline python/semantic_kernel/ai/open_ai/services/azure_chat_completion.py:55]: why allow http:// with AAD? sending AAD tokens over http will expose them to MITM

Comment (jjhenkel) [reply python/semantic_kernel/ai/open_ai/services/azure_chat_completion.py]: Good catch, I was testing at one point and swapped endpoint url and api key (and wrongly thought that, in AD contexts, the endpoint URL could look funky; just double checked and, even in AD contexts, it is indeed a normal https:// url)

Fixed!

File: python/tests/unit/ai/open_ai/services/test_azure_text_completion.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+            endpoint="",
+            api_key=api_key,
+            api_version=api_version,
+            logger=logger,
+        )
+
+
+def test_azure_text_completion_init_with_invalid_endpoint() -> None:
+    deployment_name = "test_deployment"
+    # endpoint = "https://test-endpoint.com"
```
Comment (awharrison-28) [inline python/tests/unit/ai/open_ai/services/test_azure_text_completion.py:91]:
```suggestion
    endpoint = "test-endpoint.com"
```

File: python/semantic_kernel/ai/open_ai/services/azure_text_embedding.py [hunk 5]
```diff
@@ -32,13 +57,14 @@ def __init__(
 
         self._endpoint = endpoint
         self._api_version = api_version
+        self._api_type = "azure_ad" if use_ad_auth else "azure"
```
Comment (awharrison-28) [inline python/semantic_kernel/ai/open_ai/services/azure_text_embedding.py:60]: use_ad_auth is a boolean here. Should this be "ad_auth"?

**Review by awharrison-28 — CHANGES_REQUESTED** [review]

File: python/tests/unit/ai/open_ai/services/test_azure_text_completion.py (hunk 4 above)
Comment (jjhenkel) [reply python/tests/unit/ai/open_ai/services/test_azure_text_completion.py]: Did a version of this (`endpoint = "http://test-endpoint"` and then used `endpoint=endpoint` in ctor to match other tests)

File: python/semantic_kernel/ai/open_ai/services/azure_text_embedding.py (hunk 5 above)
Comment (jjhenkel) [reply python/semantic_kernel/ai/open_ai/services/azure_text_embedding.py]: Renamed to `ad_auth`

**Review by awharrison-28 — APPROVED** [review]
LGTM and verified manually

**Review by dluc — APPROVED** [review]

**Comment by bowleri** [comment]
Where is the documentation for this use case. Providing a TokenCredential instead of an API Key does nothing. Nothing in the source code indicates a token or signed_session is being retrieved or used for authentication

**Pull request merged**